src/workerd/server/server.c++

```diff
@@ -2494,6 +2497,94 @@ kj::Own<Server::Service> Server::makeWorker(kj::StringPtr name, config::Worker::
+               "You must run workerd with `--experimental` to use this feature.");
+    // If the config has the moduleFallback option, then we are going to set up the ability
+    // to load certain modules from a fallback service. This is generally intended for local
+    // dev/testing purposes only.
+    auto& apiIsolate = isolate->getApiIsolate();
+    apiIsolate.setModuleFallbackCallback(
+        [address=kj::str(conf.getModuleFallback()), featureFlags=apiIsolate.getFeatureFlags()]
+        (jsg::Lock& js,
+         kj::StringPtr specifier,
+         jsg::CompilationObserver& observer,
+         jsg::ModuleRegistry::ResolveMethod method) mutable
+            -> kj::Maybe<jsg::ModuleRegistry::ModuleInfo> {
+      kj::Maybe<kj::String> jsonPayload;
+      {
+        // Module loading in workerd is expected to be synchronous but we need to perform
+        // an async HTTP request to the fallback service. To accomplish that we wrap the
+        // actual request in a kj::Thread, perform the GET, and drop the thread immediately
+        // so that the destructor joins the current thread (blocking it). The thread will
+        // either set the jsonPayload variable or not.
+        kj::Thread loaderThread(
+          [specifier,address=address.asPtr(),&jsonPayload, method]() mutable {
+          try {
+            const auto toStr = [](jsg::ModuleRegistry::ResolveMethod method) {
+              switch (method) {
+                case jsg::ModuleRegistry::ResolveMethod::IMPORT: return "import"_kjc;
+                case jsg::ModuleRegistry::ResolveMethod::REQUIRE: return "require"_kjc;
+              }
+              KJ_UNREACHABLE;
+            };
+
+            kj::AsyncIoContext io = kj::setupAsyncIo();
+
+            kj::HttpHeaderTable::Builder builder;
+            kj::HttpHeaderId kMethod = builder.add("resolve-method");
+            auto headerTable = builder.build();
+
+            auto addr = kj::heap<PromisedNetworkAddress>(
+                io.provider->getNetwork().parseAddress(address, 80));
+
+            auto client = kj::newHttpClient(
+                io.provider->getTimer(),
+                *headerTable,
+                *addr, {});
+
+            kj::HttpHeaders headers(*headerTable);
+            headers.set(kMethod, toStr(method));
+            headers.set(kj::HttpHeaderId::HOST, "localhost"_kj);
+
+            auto request = client->request(
+                kj::HttpMethod::GET,
+                kj::str("file://", specifier),
+                headers, kj::none);
+
+            auto resp = request.response.wait(io.waitScope);
+
+            if (resp.statusCode != 200) {
+              // Failed!
+              auto payload = resp.body->readAllText().wait(io.waitScope);
+              KJ_LOG(ERROR, "Fallback service failed to fetch module", payload);
+            } else {
+              jsonPayload = kj::str(resp.body->readAllText().wait(io.waitScope));
+            }
+          } catch (...) {
```

Comment:
Yeah, I know. I have a general preference for keeping the try/catch within the child thread here, especially since we're not actually propagating the error up but simply logging it.